rfcs/0000-c-unwind-abi.md

```diff
@@ -0,0 +1,288 @@
+boundary, either from a `panic!` "escaping" from a Rust function defined with
+`extern "C"` or by entering Rust from another language via an entrypoint
+declared with `extern "C"`, caused undefined behavior.
+
+This RFC retains most of that undefined behavior, with two exceptions:
+
+* With the `panic=unwind` runtime, `panic!` will cause an `abort` if it would
+  otherwise "escape" from a function defined with `extern "C"`.
+* Forced unwinding is safe with `extern "C"` as long as no frames with
+  destructors (i.e. `Drop` types) are unwound. This is to keep behavior of
+  `pthread_exit` and `longjmp` consistent across platforms.
+
+## Interaction with `panic=abort`
+
+If a non-forced foreign unwind would enter a Rust frame via an `extern "C
+unwind"` ABI boundary, but the Rust code is compiled with `panic=abort`, the
+unwind will be caught and the process aborted.
+
+There are some types of unwinding that are not guaranteed to cause the program
+to abort with `panic=abort`, though:
+
+* Forced unwinding: Rust provides no mechanism to catch this type of unwinding.
+  This is safe with either the `"C"` ABI or the new `"C unwind"` ABI, as long
+  as none of the unwound Rust frames contain destructors.
+* Unwinding from another language into Rust if the entrypoint to that language
+  is declared with `extern "C"` (contrary to the guidelines above): this is
+  always undefined behavior.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This table shows the behavior of an unwinding operation reaching each type of
+ABI boundary (function declaration or definition). "UB" stands for undefined
+behavior. `"C"`-like ABIs are `"C"` itself but also related ABIs such as
+`"system"`.
+
+| panic runtime  | ABI          | `panic`-unwind                        | Unforced foreign unwind |
+| -------------- | ------------ | ------------------------------------- | ----------------------- |
+| `panic=unwind` | `"C unwind"` | unwind                                | unwind                  |
+| `panic=unwind` | `"C"`-like   | abort                                 | UB                      |
+| `panic=abort`  | `"C unwind"` | `panic!` aborts                       | abort                   |
+| `panic=abort`  | `"C"`-like   | `panic!` aborts (no unwinding occurs) | UB                      |
+
+Regardless of the panic runtime, ABI, or platform, the interaction of Rust
+frames with C functions that deallocate frames (i.e. functions that may use
+forced unwinding on specific platforms) is specified as follows:
+
+* **When deallocating Rust frames without destructors:** frames are safely
+  deallocated; no undefined behavior
+* **When deallocating Rust frames with destructors:** undefined behavior
+
+No subtype relationship is defined between functions or function pointers using
+different ABIs. This RFC also does not define coercions between `"C"` and
+`"C unwind"`.
```

Comment:
Maybe we should include a coercion from "C" to "C unwind"

Reply:
Hm -- for simplicity, for now, I'd rather not. I could probably be persuaded otherwise, but I think we'd have a better idea of whether this would actually be substantially beneficial after `"C unwind"` has been available to the general public for a while.